seastar/src/file.rs

```diff
@@ -0,0 +1,298 @@
+#[derive(Clone)]
+pub struct OpenOptions {
+    read: bool,
+    write: bool,
+    create: bool,
+}
+
+impl Default for OpenOptions {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl OpenOptions {
+    pub fn new() -> Self {
+        Self {
+            read: false,
+            write: false,
+            create: false,
+        }
+    }
+
+    pub fn read(&mut self, flag: bool) -> &mut Self {
+        self.read = flag;
+        self
+    }
+
+    pub fn write(&mut self, flag: bool) -> &mut Self {
+        self.write = flag;
+        self
+    }
+
+    pub fn create(&mut self, flag: bool) -> &mut Self {
+        self.create = flag;
+        self
+    }
+
+    pub fn get_read(&self) -> bool {
+        self.read
+    }
+
+    pub fn get_write(&self) -> bool {
+        self.write
+    }
+
+    pub fn get_create(&self) -> bool {
+        self.create
+    }
+
+    pub async fn open<P: AsRef<Path>>(&self, path: P) -> io::Result<File> {
+        File::new(&self.clone(), path.as_ref()).await
+    }
+}
+
+pub struct File {
+    inner: UniquePtr<file_t>,
+}
+
+impl File {
+    // Creates a new file with `opts` OpenOptions and `path` path to a file.
+    //
+    // Returns a file.
+    pub async fn new(opts: &OpenOptions, path: &Path) -> io::Result<File> {
+        assert_runtime_is_running();
+        let mut f_ptr = UniquePtr::null();
+        let name = path.to_str().unwrap();
+        let res = open_dma(&mut f_ptr, name, opts).await;
+        match res {
+            Ok(_) => Ok(File { inner: f_ptr }),
+            Err(_) => Err(io::Error::new(io::ErrorKind::Other, "No read permission")),
+        }
+    }
+
+    // Read some bytes at given position.
+    //
+    // The function does not overwrite existing data in the DmaBuffer,
+    // but rather reads the data into the free capacity of the internal Vec.
+    //
+    // Returns the number of bytes read and the original buffer.
+    pub async fn read_dma(
+        &self,
+        buffer: DmaBuffer,
+        pos: u64,
+    ) -> Result<(usize, DmaBuffer), io::Error> {
+        assert_runtime_is_running();
+        let buf_ptr = buffer.chunks.as_ptr() as *mut u8;
+        let size = buffer.chunks.len() as u64;
+        unsafe {
+            let fut = read_dma(&self.inner, buf_ptr, size * 4096, pos);
+            match fut.await {
+                Ok(res) => Ok((res as usize, buffer)),
+                Err(e) => Err(io::Error::new(io::ErrorKind::Other, e)),
+            }
+        }
+    }
+
+    // Writes some bytes at given position.
+    //
+    // Returns (bytes_written, original_buffer) on success.
+    pub async fn write_dma(
+        &self,
+        buffer: DmaBuffer,
+        pos: u64,
+    ) -> Result<(usize, DmaBuffer), io::Error> {
+        assert_runtime_is_running();
+        let buf_ptr = buffer.chunks.as_ptr() as *mut u8;
+        let size = buffer.chunks.len() as u64;
+        unsafe {
+            let fut = write_dma(&self.inner, buf_ptr, size * 4096, pos);
```

Comment:
This is may be just a subjective opinion, but I feel this multiplication should me move to `size` declaration.